src/responder.ts

```diff
@@ -51,182 +61,357 @@ enum ResponderEvent {
-     * @param appointmentId The id of the Appointment this object is responding to.
-     * @param ethereumResponse The IEthereumResponse containing what needs to be submitted.
-     */
-    constructor(
-        readonly signer: ethers.Signer,
-        public readonly appointmentId: string,
-        public readonly ethereumResponse: IEthereumResponse
-    ) {
+
+    // implementations should query the provider (or a service) to figure out the appropriate gas price
+    protected gasPrice = new ethers.utils.BigNumber(21000000000);
+
+    constructor(public readonly signer: ethers.Signer) {
         super();
     }
 
-    protected submitStateFunction(): Promise<TransactionResponse> {
+    /**
+     * Creates the transaction request to be sent to handle the response in `resposeData`.
+     *
+     * @param responseData the response data used to create the transaction
+     * @param nonce The nonce to be used. If `null`, the provider will set the nonce. It is recommended to explicitly provide a nonce,
+     *              especially if the same wallet might send multiple transactions concurrently or in a short span.
+     */
+    protected prepareTransactionRequest(responseData: IEthereumResponseData, nonce: number = null): ethers.providers.TransactionRequest {
         // form the interface so that we can serialise the args and the function name
-        const abiInterface = new ethers.utils.Interface(this.ethereumResponse.contractAbi);
-        const data = abiInterface.functions[this.ethereumResponse.functionName].encode(this.ethereumResponse.functionArgs);
+        const abiInterface = new ethers.utils.Interface(responseData.contractAbi);
+        const data = abiInterface.functions[responseData.functionName].encode(responseData.functionArgs);
         // now create a transaction, specifying possible oher variables
-        const transactionRequest = {
-            to: this.ethereumResponse.contractAddress,
+        return {
+            to: responseData.contractAddress,
             gasLimit: EthereumResponder.GAS_LIMIT,
-            // nonce: 0,
-            gasPrice: 21000000000, // TODO: choose an appropriate gas price
+            nonce: nonce,
+            gasPrice: this.gasPrice,
             data: data
         };
-
-        // execute the transaction
-        return this.signer.sendTransaction(transactionRequest);
     }
+
+    /**
+    * @param appointmentId The id of the Appointment this object is responding to.
+    * @param response The IEthereumResponse containing what needs to be submitted.
+    */
+    public abstract startResponse(appointmentId: string, responseData: IEthereumResponseData): void;
 }
 
 
 /* CONCRETE RESPONDER IMPLEMENTATIONS */
 
+/**
+ * A gas policy implements the strategy for the choice of the gas price for subsequent attempts at submitting a transaction.
+ */
+export interface GasPolicy {
```

Comment:
I've been starting interface names with a capital I. It's just a convention no real reason, dyou have any preferences on that?

Reply:
I guess it makes sense to keep the capital I standard for interfaces; initially I had an abstract class instead. I'll change it.

Reply:
Done in 352fc270d00d7838ef20e737c027295a3bb1c97d.